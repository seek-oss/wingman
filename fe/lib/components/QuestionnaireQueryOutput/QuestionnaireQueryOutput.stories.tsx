import 'braid-design-system/reset';
import React, { ComponentProps } from 'react';

import type { FormComponent } from '../../private/questionnaires/types';
import {
  BraidArgs,
  defaultArgTypes,
  defaultArgs,
} from '../../storybook/controls';
import { BraidStorybookProvider } from '../../storybook/decorators';

import { QuestionnaireQueryOutput as Component } from './QuestionnaireQueryOutput';

const sampleQuestions: FormComponent[] = [
  {
    componentTypeCode: 'Question',
    questionHtml: 'Tell us about yourself',
    responseTypeCode: 'FreeText',
    value: '04e671b1-166d-416d-a346-4b560ad13a51',
  },
  {
    componentTypeCode: 'Question',
    questionHtml: 'Can you drive?',
    responseChoice: [
      { preferredIndicator: true, text: 'Yes', value: 'yes' },
      { preferredIndicator: false, text: 'No', value: 'no' },
    ],
    responseTypeCode: 'SingleSelect',
    value: '76250b99-dfc7-47a3-8418-1fc6bee64589',
  },
  {
    componentTypeCode: 'PrivacyConsent',
    privacyPolicyUrl: {
      url: 'https://www.seek.com.au/my-privacy/',
    },
    value: 'cdb93e19-6caf-4852-854c-3088787f234c',
    descriptionHtml: 'Do you agree to the privacy policy?',
  },
];

export default {
  args: {
    braidThemeName: defaultArgs.braidThemeName,
    components: sampleQuestions,
    hirerId: 'seekAnzPublicTest:organization:seek:93WyyF1h',
  },
  argTypes: {
    braidThemeName: defaultArgTypes.braidThemeName,
  },
  component: Component,
  title: 'Job Posting/Questionnaires/QuestionnaireQueryOutput',
};

type Args = ComponentProps<typeof Component> & BraidArgs;

export const QuestionnaireQueryOutput = ({ braidThemeName, ...args }: Args) => (
  <BraidStorybookProvider braidThemeName={braidThemeName}>
    <Component {...args} />
  </BraidStorybookProvider>
);
QuestionnaireQueryOutput.storyName = 'QuestionnaireQueryOutput';
