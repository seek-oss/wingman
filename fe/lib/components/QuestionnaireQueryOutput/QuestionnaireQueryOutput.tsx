import {
  RadioGroup,
  RadioItem,
  Stack,
  Text,
  TextLink,
} from 'braid-design-system';
import React, { useState } from 'react';
import { CodeBlock } from 'scoobie';

import { MAX_NUMBER_OF_COMPONENTS } from '../../private/questionnaires/constants';
import type {
  FormComponent,
  PrivacyConsent,
  Question,
} from '../../private/questionnaires/types';

type MutationQuestionnaireComponents =
  | {
      componentTypeCode: 'Question';
      question: Question;
    }
  | {
      componentTypeCode: 'PrivacyConsent';
      privacyConsent: PrivacyConsent;
    };

export interface QuestionnaireCreateInput {
  variables: {
    input: {
      applicationQuestionnaire: {
        hirerId: string;
        components: MutationQuestionnaireComponents[];
      };
    };
  };
  valid: boolean;
}

const graphqlMutation = `
mutation($input: CreateApplicationQuestionnaireInput!) {
  createApplicationQuestionnaire(input: $input) {
    applicationQuestionnaire {
      id {
        value
      }
    }
  }
}`.trim();

const validateComponents = (components: FormComponent[]) => {
  if (components.length === 0 || components.length > MAX_NUMBER_OF_COMPONENTS) {
    return false;
  }

  const privacyConsentComponents = components.filter(
    (fc): fc is PrivacyConsent => fc.componentTypeCode === 'PrivacyConsent',
  );

  if (privacyConsentComponents.length > 1) {
    return false;
  }

  return true;
};

export const convertComponentsToMutationVariables = (
  components: FormComponent[],
  hirerId: string,
): QuestionnaireCreateInput => {
  const questionnaireComponents: MutationQuestionnaireComponents[] =
    components.map((component) =>
      component.componentTypeCode === 'Question'
        ? {
            componentTypeCode: 'Question',
            question: component,
          }
        : {
            componentTypeCode: 'PrivacyConsent',
            privacyConsent: component,
          },
    );

  return {
    variables: {
      input: {
        applicationQuestionnaire: {
          hirerId,
          components: questionnaireComponents,
        },
      },
    },
    valid: validateComponents(components),
  };
};

const renderVariablesPane = (components: FormComponent[], hirerId: string) => {
  if (components.length > 0) {
    const { variables } = convertComponentsToMutationVariables(
      components,
      hirerId,
    );

    const variablesString = JSON.stringify(variables, null, 2);

    return (
      <Stack space="medium">
        <Text>For the variables of the mutation, use:</Text>

        <CodeBlock language="json">{variablesString}</CodeBlock>
      </Stack>
    );
  }

  return <Text>You need to add some questions first.</Text>;
};

interface QuestionnaireQueryOutputProps {
  components: FormComponent[];
  hirerId: string;
}

export const QuestionnaireQueryOutput = ({
  components,
  hirerId,
}: QuestionnaireQueryOutputProps) => {
  type VisiblePane = 'variables' | 'mutation';
  const [visiblePane, setVisiblePane] = useState<VisiblePane>('variables');

  return (
    <Stack space="large">
      <RadioGroup
        aria-label="GraphQL query"
        id="graphql-query-renderer-pane"
        onChange={(event) =>
          setVisiblePane(event.currentTarget.value as VisiblePane)
        }
        value={visiblePane}
      >
        <RadioItem label="Variables" value="variables" />
        <RadioItem label="Mutation" value="mutation" />
      </RadioGroup>

      {visiblePane === 'mutation' && (
        <Stack space="medium">
          <Text>
            To generate the form you’ve built, you’ll need to send a mutation to
            the{' '}
            <TextLink href="https://graphql.seek.com/graphql" target="_blank">
              SEEK API
            </TextLink>
            . It will look something like this, although the response fields you
            require are up to you.
          </Text>

          <CodeBlock language="graphql">{graphqlMutation}</CodeBlock>
        </Stack>
      )}

      {visiblePane === 'variables' && renderVariablesPane(components, hirerId)}
    </Stack>
  );
};
