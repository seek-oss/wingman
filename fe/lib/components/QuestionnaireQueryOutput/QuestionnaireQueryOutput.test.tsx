import type { FormComponent } from '../../private/questionnaires/types';

import { convertComponentsToMutationVariables } from './QuestionnaireQueryOutput';

describe('convertComponentsToMutationVariables', () => {
  describe('when converting components to mutation variables', () => {
    it('should return as an application questionnaire and return as valid if it meets the conditions', () => {
      const formComponents: FormComponent[] = [
        {
          componentTypeCode: 'Question',
          questionHtml: 'Tell us about yourself',
          responseTypeCode: 'FreeText',
          value: '04e671b1-166d-416d-a346-4b560ad13a51',
        },
        {
          componentTypeCode: 'Question',
          questionHtml: 'Can you drive?',
          responseChoice: [
            { preferredIndicator: true, text: 'Yes', value: 'yes' },
            { preferredIndicator: false, text: 'No', value: 'no' },
          ],
          responseTypeCode: 'SingleSelect',
          value: '76250b99-dfc7-47a3-8418-1fc6bee64589',
        },
        {
          componentTypeCode: 'PrivacyConsent',
          privacyPolicyUrl: {
            url: 'https://www.seek.com.au/my-privacy/',
          },
          value: 'cdb93e19-6caf-4852-854c-3088787f234c',
          descriptionHtml: 'Do you agree to the privacy policy?',
        },
      ];
      expect(
        convertComponentsToMutationVariables(formComponents, '123'),
      ).toMatchSnapshot();
    });

    it('should return as invalid if there arent any components', () => {
      const formComponents: FormComponent[] = [];
      expect(
        convertComponentsToMutationVariables(formComponents, '123').valid,
      ).toBe(false);
    });

    it('should return as invalid if there are too many components', () => {
      const formComponents: FormComponent[] = Array(101).fill({
        componentTypeCode: 'Question',
        questionHtml: 'Tell us about yourself',
        responseTypeCode: 'FreeText',
        value: '04e671b1-166d-416d-a346-4b560ad13a51',
      });
      expect(
        convertComponentsToMutationVariables(formComponents, '123').valid,
      ).toBe(false);
    });

    it('should return as invalid if there are too many PrivacyConsent components', () => {
      const formComponents: FormComponent[] = Array(2).fill({
        componentTypeCode: 'PrivacyConsent',
        privacyPolicyUrl: {
          url: 'https://www.seek.com.au/my-privacy/',
        },
        value: 'cdb93e19-6caf-4852-854c-3088787f234c',
        descriptionHtml: 'Do you agree to the privacy policy?',
      });
      expect(
        convertComponentsToMutationVariables(formComponents, '123').valid,
      ).toBe(false);
    });
  });
});
