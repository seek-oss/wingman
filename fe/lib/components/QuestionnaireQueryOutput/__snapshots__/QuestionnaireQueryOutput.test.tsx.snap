// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertComponentsToMutationVariables when converting components to mutation variables should return as an application questionnaire and return as valid if it meets the conditions 1`] = `
{
  "valid": true,
  "variables": {
    "input": {
      "applicationQuestionnaire": {
        "components": [
          {
            "componentTypeCode": "Question",
            "question": {
              "componentTypeCode": "Question",
              "questionHtml": "Tell us about yourself",
              "responseTypeCode": "FreeText",
              "value": "04e671b1-166d-416d-a346-4b560ad13a51",
            },
          },
          {
            "componentTypeCode": "Question",
            "question": {
              "componentTypeCode": "Question",
              "questionHtml": "Can you drive?",
              "responseChoice": [
                {
                  "preferredIndicator": true,
                  "text": "Yes",
                  "value": "yes",
                },
                {
                  "preferredIndicator": false,
                  "text": "No",
                  "value": "no",
                },
              ],
              "responseTypeCode": "SingleSelect",
              "value": "76250b99-dfc7-47a3-8418-1fc6bee64589",
            },
          },
          {
            "componentTypeCode": "PrivacyConsent",
            "privacyConsent": {
              "componentTypeCode": "PrivacyConsent",
              "descriptionHtml": "Do you agree to the privacy policy?",
              "privacyPolicyUrl": {
                "url": "https://www.seek.com.au/my-privacy/",
              },
              "value": "cdb93e19-6caf-4852-854c-3088787f234c",
            },
          },
        ],
        "hirerId": "123",
      },
    },
  },
}
`;
